[Autotuner] Improve python package management

luarss opened the pull request:
- uses `uv`, a fast rust based package manager for python.
- handles dependencies from python 3.8 - 3.10

Fixes https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues/2888

luarss committed:
add uv for package management

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix pyproject to include py3.10

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
enforce ray <= 2.23, local_dir option is deprecated from ray 2.24 onwards

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on tools/AutoTuner/pyproject.toml:
```diff
@@ -1,26 +1,37 @@
-requires-python = ">= 3.8"
-dynamic = ["dependencies", "optional-dependencies"]
+dependencies = [
+    "ray[default,tune] >= 2.10, <= 2.23",
```
To close #2859 we need at least `2.40`. We can continue with the PR, but this will not be enough to close the issue.

luarss replied:
Removed the issue link. I will fix some ray compatibility issues, marking this as a draft

luarss committed:
fix distributed relative path bug (ray 2.10 onwards starts ray in /tmp/ray)

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix uv lock, fix log_path for new ray (changed working directory to /tmp instead of local)

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix file path in plot, refactor openroad util to populate paths correctly

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix openroad_distributed

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
fix ref_file_check path

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
upgrade urllib

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
test-resume: increase number of samples for iterations 2 to work more reliably
- currently under gp/dp constraint -> 0.625 of the configurations are valid. if we use 5 samples -> only 0.625*0.625*5 ~ 1.9 samples will be valid.
- if we use 8 samples ~ 3.1 samples will be valid.

Signed-off-by: Jack Luar <jluar@precisioninno.com>

vvbandeira commented on tools/AutoTuner/src/autotuner/distributed.py:
```diff
@@ -87,9 +87,7 @@ def setup(self, config):
-        # Run by Ray in directory specified by `local_dir`
-        repo_dir = os.getcwd() + "/../" * 6
-        self.repo_dir = os.path.abspath(repo_dir)
+        self.repo_dir = os.path.abspath(LOCAL_DIR + "/../" * 4)
```
Update the comment above.

vvbandeira commented on tools/AutoTuner/src/autotuner/utils/plot.py:
```diff
@@ -41,7 +41,9 @@ def load_dir(dir: str) -> pd.DataFrame:
     for params_fname in glob.glob(f"{dir}/*/params.json"):
-        metrics_fname = params_fname.replace("params.json", "metrics.json")
+        metrics_fname = params_fname.replace("params.json", "metrics.json").replace(
+            "ray", "or-0"
```
Why do we need to replace `ray` with `or-0`?

luarss replied:
In latest Ray versions this is an example file structure (`or` is replaced with `or-0`)

```
autotuner-best-5dfabe02.json               searcher-state-2025-02-18_17-10-34.pkl
experiment_state-2025-02-18_17-10-34.json  variant-AutoTunerBase-5dfabe02-or-0
search_gen_state-2025-02-18_17-10-34.json  variant-AutoTunerBase-5dfabe02-ray
```

luarss replied:
https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Public/job/PR-2859-head/lastSuccessfulBuild/artifact/flow/logs/asap7/gcd/tune-df80c9dd/

vvbandeira commented on docs/user/InstructionsForAutoTuner.md:
```diff
@@ -232,6 +232,11 @@ python3 ./tools/AutoTuner/test/smoke_test_tune.py
 
+## Developer reference
+
+1. The [`uv`](https://docs.astral.sh/uv) package manager is used for handling different package requirements in AutoTuner.
```
What problem `uv` resolves that `pip` or `venv` do not?

luarss replied:
- Fastest python package manager/resolver
- It also acts as a catch-all for `pip`, `venv` and additionally `pip-tools` - which can manage python dependencies across different operating systems and versions.

vvbandeira replied:
`pip` and `venv` are cross OS, right? And if we are committing the requirements file, this feature of supporting multiple versions (of Python?) will be lost.

luarss replied:
The committed requirements file actually is the solved deps for these versions, if not an error will be reported during uv compilation.

vvbandeira commented on tools/AutoTuner/Makefile:
```diff
@@ -0,0 +1,7 @@
+	@rm -f requirements.txt
+	@rm -f requirements-test.txt
+	@uv pip compile --output-file=requirements.txt pyproject.toml --upgrade
+	@uv pip compile --output-file=requirements-dev.txt pyproject.toml --extra dev  --upgrade
```
If we are generating the files, why keep them in git?

luarss replied:
For faster dependency resolving. The idea is to first store the solved dependencies in `uv.lock` and `requirements*.txt` file and next time, we can just install directly from these files

luarss replied:
This also has a side benefit of locking the python versions installed.

vvbandeira replied:
I am still not sold on this. What is the process exactly? We update `pyproject.toml`, then use `uv` to generate `uv.lock` and `requirements.txt`, and finally use `pip` to install the `requirements.txt`? Why not just write `requirements.txt`? This looks like many extra steps that need to be tracked manually.

If we are using pip to install the dependencies at the end, looks to me that all the extra info that `uv` saves in the lock file is moot.

vvbandeira replied:
Please provide an example of how the files are generated, in order, and used.
Also, provide another example to showcase a situation where having `uv` will be beneficial.

luarss replied:
The idea i was aiming for is to periodically run the uv requirements updater (ci or manually) and if it passes test, merge the updated requirements. This step is supposed to be dev-only, so users need not worry. 

You are right that uv.lock file is not used in production - that is just to minimise complexity in Dockerfile. We use vanilla pip install on the compiled requirements.txt file.

edit: No other benefits for now, since we are not using uv.lock in production

vvbandeira commented on tools/AutoTuner/Makefile:
```diff
@@ -0,0 +1,7 @@
+	@rm -f requirements-test.txt
+	@uv pip compile --output-file=requirements.txt pyproject.toml --upgrade
+	@uv pip compile --output-file=requirements-dev.txt pyproject.toml --extra dev  --upgrade
+	@uv sync --all-extras --inexact
```
What does the `inexact` flag do? Will this mean that we will have a non-exact version of packages?

luarss replied:
Removing.

luarss committed:
remove inexact

Signed-off-by: Jack Luar <jluar@precisioninno.com>

luarss committed:
update comment

Signed-off-by: Jack Luar <jluar@precisioninno.com>
